consensus/src/processes/pruning_proof/build.rs

```diff
@@ -173,42 +198,100 @@ impl PruningProofManager {
                     .map_err(|err| format!("level + 1: {}, err: {}", level + 1, err))
                     .unwrap();
                 Some(block_at_depth_m_at_next_level)
             } else {
                 None
             };
-            let (store, selected_tip, root) = self
-                .find_sufficiently_deep_level_root(pp_header, level, required_block, temp_db.clone())
-                .unwrap_or_else(|_| panic!("find_sufficient_root failed for level {level}"));
-            ghostdag_stores[level_usize] = Some(store);
-            selected_tip_by_level[level_usize] = Some(selected_tip);
-            root_by_level[level_usize] = Some(root);
+            level_proof_stores_vec[level_usize] = Some(
+                self.find_sufficiently_deep_level_root(pp_header, level, required_block, temp_db.clone())
+                    .unwrap_or_else(|_| panic!("find_sufficient_root failed for level {level}")),
+            );
+        }
+
+        let (transient_ghostdag_stores, transient_relations_stores, selected_tip_by_level, roots_by_level) =
+            level_proof_stores_vec.into_iter().map(Option::unwrap).multiunzip();
+
+        MultiLevelProofContext { transient_ghostdag_stores, transient_relations_stores, selected_tip_by_level, roots_by_level }
+    }
+
+    fn populate_relation_store_at_level(
+        &self,
+        temp_db: Arc<DB>,
+        tip: Hash,
+        root: Hash,
+        level: BlockLevel,
+        try_number: u8,
+    ) -> Arc<RwLock<DbRelationsStore>> {
+        //TODO(relaxed): currently we rebuild the relation store in its entirety for each try.
+        // A more efficient implementation could instead extend the store constructed in the previous try
+        //TODO: Remove sanity checks when convinced code is stable and correct
+        let mut queue = VecDeque::new();
+        let mut visited = BlockHashSet::new();
+        queue.push_back(tip);
+        let cache_policy = CachePolicy::Count(2 * self.pruning_proof_m as usize);
+        let level_relation_store =
+            Arc::new(RwLock::new(DbRelationsStore::new_temp(temp_db.clone(), level, try_number, cache_policy, cache_policy)));
+        let mut relations_write = level_relation_store.write();
+
+        relations_write.insert(ORIGIN, BlockHashes::new(vec![])).unwrap();
+
+        // sanity check
+        assert!(self.reachability_service.is_dag_ancestor_of(root, tip));
+        while let Some(hash) = queue.pop_front() {
+            if !visited.insert(hash) {
+                continue;
+            }
+            if hash == ORIGIN {
+                continue;
+            }
+            let header = self.headers_store.get_header(hash).unwrap();
+            let parents = Arc::new(
+                self.parents_manager
+                    .parents_at_level(&header, level)
+                    .iter()
+                    .copied()
+                    .filter(|&p| self.reachability_service.is_dag_ancestor_of_result(root, p).unwrap_or(false))
```

Comment:
```suggestion
                    .filter(|&p| self.reachability_service.is_dag_ancestor_of_result(root, p).unwrap_option().unwrap_or(false))
```
The current `unwrap_or` swallows all DB/IO errors as well, whereas you only want to ignore a missing entry and treat it as false